libraries/AP_HAL_Linux/GPIO_RPI.cpp

```diff
@@ -19,85 +20,186 @@
-#define GPIO_SET_HIGH       *(_gpio+7)  // sets   bits which are 1
-#define GPIO_SET_LOW        *(_gpio+10) // clears bits which are 1
-#define GPIO_GET(g)         (*(_gpio+13)&(1<<g)) // 0 if LOW, (1<<g) if HIGH
 #define GPIO_RPI_MAX_NUMBER_PINS 32
 
 using namespace Linux;
 
 extern const AP_HAL::HAL& hal;
 
+// Range based in the first memory address of the first register and the last memory addres
+// for the GPIO section (0x7E20'00B4 - 0x7E20'0000).
+const uint8_t GPIO_RPI::_gpio_registers_memory_range = 0xB4;
+const char* GPIO_RPI::_system_memory_device_path = "/dev/mem";
+
 GPIO_RPI::GPIO_RPI()
 {
 }
 
+void GPIO_RPI::set_gpio_mode_alt(int pin, int alternative)
+{
+    // Each register can contain 10 pins
+    const uint8_t pins_per_register = 10;
+    // Calculates the position of the 3 bit mask in the 32 bits register
+    const uint8_t tree_bits_position_in_register = (pin%pins_per_register)*3;
+    /** Creates a mask that enables the alternative function based in the following logic:
+     *
+     * | Alternative Function | 3 bits value |
+     * |:--------------------:|:------------:|
+     * |      Function 0      |     0b100    |
+     * |      Function 1      |     0b101    |
+     * |      Function 2      |     0b110    |
+     * |      Function 3      |     0b111    |
+     * |      Function 4      |     0b011    |
+     * |      Function 5      |     0b010    |
+     */
+    const uint8_t alternative_value =
+        (alternative < 4 ? (alternative + 4) : (alternative == 4 ? 3 : 2));
+    // 0b00'000'000'000'000'000'000'ALT'000'000'000 enables alternative for the 4th pin
+    alternative = alternative_value << tree_bits_position_in_register;
+    const uint32_t mask = static_cast<uint32_t>(alternative) << tree_bits_position_in_register;
+    // Apply mask
+    _gpio[pin / pins_per_register] |= mask;
+}
+
+void GPIO_RPI::set_gpio_mode_in(int pin)
+{
+    // Each register can contain 10 pins
+    const uint8_t pins_per_register = 10;
+    // Calculates the position of the 3 bit mask in the 32 bits register
+    const uint8_t tree_bits_position_in_register = (pin%pins_per_register)*3;
+    // Create a mask that only removes the bits in this specific GPIO pin, E.g:
+    // 0b11'111'111'111'111'111'111'000'111'111'111 for the 4th pin
+    const uint32_t mask = ~(0b111<<tree_bits_position_in_register);
+    // Apply mask
+    _gpio[pin / pins_per_register] &= mask;
+}
+
+void GPIO_RPI::set_gpio_mode_out(int pin)
+{
+    // Each register can contain 10 pins
+    const uint8_t pins_per_register = 10;
+    // Calculates the position of the 3 bit mask in the 32 bits register
+    const uint8_t tree_bits_position_in_register = (pin%pins_per_register)*3;
+    // Create a mask that enable the bit the sets output functionality
+    // 0b00'000'000'000'000'000'000'001'000'000'000 enables output for the 4th pin
+    const uint32_t mask = 0b001 << tree_bits_position_in_register;
+    // Apply mask
+    _gpio[pin / pins_per_register] |= mask;
+}
```

Comment:
I think we have lots of duplication in these functions... example: the only difference in set_gpio_mode_in()
and set_gpio_mode_out() is just the value we are writing and that in one case we OR and the other we AND. We could improve that on top, but there is another issue:

There is a problem in doing just a & or |... if you set it to mode_in, and then to mode_out it won't work since `0x7 | 0x1 == 0x7`.

In all of these case you should be using a RMW. First you have to get the value from the hardware out in a local var, modify, and then write it back.

```c++
// mode = 0b001 or 0b111, depending on the func, then:
value = _gpio[pin / pins_per_register];
value &= ~tree_bits_position_in_register;
value |= mode;
_gpio[pin / pins_per_register] = mode;
```

Reply:
Nice catch!